### DISCUSSION TITLE
Guide for making contributions?

### DISCUSSION THREAD

**Discussion by bakhtos** [tensorly/tensorly#541] [General]
Hello,

I will be using tensorly for my PhD research in the near future to study Temporal Networks a lot, and I guess I will end up making contributions to tensorly as well (for example, some methods that I plan to use actually require good-old matrix algebra, which is not readily available in tensorly as far as I understand, so I could try contributing it).
So I wanted to ask - what is the "correct" way to introduce functionality to tensorly? Must I cover several backends myself before contributing? Should all new functions be first put into an experimental module? Is there a guideline to how tests should be written? The Contribution section of the README only mentions running the syntax checker.
I think a project such as this covering several backends should have some pipeline for how new functionality should be contributed. Sorry if I missed it somewhere, please point me in the right direction or we could discuss it here if it does not exist yet.

Thank you, 
Alexander

**Comment by yngvem** [comment]
Great that you want to contribute to TensorLy! Unfortunately, our contribution guides are a bit lacking, but if you submit a PR, then we'll be happy to help you get it ready for completion. It might be a good idea to write an issue first though and just state that you'll work on it, which can prevent multiple developers working on the same thing :)

Regarding functionality: to be merged into TensorLy, you should have code that works on all backends with tests as well. However, you don't need to have a 100% complete solution before you submit something. We'll try to help you as best as we can! We can sometimes be a bit slow, just remember that we're essentially all either volunteering time to the project, or some might be able to get some work on TensorLy in as a part of a different research project.

About tests: You're pretty much free to make them however you'd like as long as you use PyTest (here's a tutorial: https://realpython.com/pytest-python-testing/). 

(This is all just my opinion though so please correct me if I'm wrong @JeanKossaifi )
